### ISSUE TITLE
cmd/compile: implement type-based alias analysis

### ISSUE THREAD

**Issue by andreybokhanko** [golang/go#70488]
### Go version

go version devel go1.24-3ca78afb3b Mon Nov 18 04:56:52 2024 +0000 linux/arm64

### Output of `go env` in your module/workspace:

```shell
AR='ar'
CC='gcc'
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_ENABLED='1'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
CXX='g++'
GCCGO='gccgo'
GO111MODULE=''
GOARCH='arm64'
GOARM64='v8.0'
GOAUTH='netrc'
GOBIN=''
GOCACHE='/home/abokhanko/.cache/go-build'
GODEBUG=''
GOENV='/home/abokhanko/.config/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOGCCFLAGS='-fPIC -pthread -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=/tmp/go-build625133068=/tmp/go-build -gno-record-gcc-switches'
GOHOSTARCH='arm64'
GOHOSTOS='linux'
GOINSECURE=''
GOMOD='/dev/null'
GOMODCACHE='/home/abokhanko/go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='linux'
GOPATH='/home/abokhanko/go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/home/abokhanko/goroot'
GOSUMDB='sum.golang.org'
GOTELEMETRY='local'
GOTELEMETRYDIR='/home/abokhanko/.config/go/telemetry'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/home/abokhanko/goroot/pkg/tool/linux_arm64'
GOVCS=''
GOVERSION='devel go1.24-3ca78afb3b Mon Nov 18 04:56:52 2024 +0000'
GOWORK=''
PKG_CONFIG='pkg-config'
```

### What did you do?

Test case:

```
package test

func Foo(x **int, y *int) int {
  *y = 10
  *x = nil
  return *y + 20
}
```
Compiled with `go build -a -gcflags='-d ssa/all/dump=Foo' test.go`.

### What did you see happen?

Read of `*y` memory location and thus, the subsequent addition of a constant are not eliminated:

```
$ cat Foo_51__trim.dump
...
    (+6) v17 = MOVDload <int> v8 v15 : R1
    (6) v19 = ADDconst <int> [20] v17 : R0
    (6) v20 = MakeResult <int,mem> v19 v15 : <>
```

### What did you expect to see?

Read of `*y` eliminated; subsequent addition of two constant values (`10`, which is the value of `*y` and `20`) folded.

**Comment by andreybokhanko** [comment]
The reason CSE or other SSA optimizations didn’t eliminated the read of `*y` is due to presence of a write to `*x`, that `ssa.disjoint` can’t disambiguate with `*y`.

However, this can be done based on type aliasing rules. My reading of the rules (https://pkg.go.dev/unsafe#Pointer) is that no non-pointer type (like `int`, in case of `*y`) can alias with a pointer type (`*int` in case of `*x`).

**Comment by randall77** [comment]
You'll want to read through https://go-review.googlesource.com/c/go/+/551381 first.
TL;DR type-based alias analysis isn't safe in Go.

I think possibly your ptr vs nonptr rule might work, but not the rest.

**Comment by andreybokhanko** [comment]
> You'll want to read through https://go-review.googlesource.com/c/go/+/551381 first. TL;DR type-based alias analysis isn't safe in Go.
> 
> I think possibly your ptr vs nonptr rule might work, but not the rest.

@randall77 thanks for pointing at this CL; I wasn't aware of it!

I read through your comments; I believe they are already handled in my code. Essentially, what my patch does is what you proposed in your [comment to CL551381](https://go-review.googlesource.com/c/go/+/551381/2#message-82930368363fc3a0c1edbe56cf03b0a39b64ba3d):

> Also, TBAA isn't completely pointless. Pointers like **byte and *byte are known not to alias based on their type, as there is no way to alias them using rule 1 described in the unsafe package (which admittedly depends on defining "equivalent memory layout" which is mentioned there but not really defined, but in any reasonable definition would imply that byte and *byte have different layouts). I'm not sure if those cases would help much in practice.

The only "extra" is `IsNonaliasable` function, that claims that types `TNIL`, `TSTRING`, `TCHAN` and `TMAP` can only alias with themselves.

I tested this patch on `go build` itself and on `sweet` benchmark -- it triggers several thousand times (unfortunately, with very little optimization effect -- perhaps due to lack of more advanced optimizations that require more advanced analysis to give a tangible effect -- a kind of chicken and egg problem) and there are no fails.

Andrey

**Comment by ianlancetaylor** [comment]
> The only "extra" is `IsNonaliasable` function, that claims that types TNIL, TSTRING, TCHAN and TMAP can only alias with themselves.

That doesn't sound right.  For example, here is a program that stores both channel and map values in the same memory location.  This is a dreadful program but it's valid Go.

```Go
package main

import (
	"fmt"
	"unsafe"
)

type union struct {
	f chan int // sometimes this is a map
}

func (u *union) setChan(c chan int) {
	u.f = c
}

func (u *union) getChan() chan int {
	return u.f
}

func (u *union) setMap(m map[int]bool) {
	u.f = *(*chan int)(unsafe.Pointer(&m))
}

func (u *union) getMap() map[int]bool {
	return *(*map[int]bool)(unsafe.Pointer(&u.f))
}

func main() {
	u := union{f: make(chan int)}
	u.setChan(make(chan int, 1))
	close(u.getChan())
	u.setMap(map[int]bool{1: true})
	fmt.Println(u.getMap()[1])
}
```

**Comment by ianlancetaylor** [comment]
map and chan values are pointers, and as such have equivalent memory layout.  The structs you mention are what those pointers point to.  My sample program above is just changing pointers.

I agree that in the current implementation a pointer and a non-pointer can not share the same memory, and that it follows that changing memory using a `**int` can't affect the memory pointed to by a `*int`.

**Comment by andreybokhanko** [comment]
> map and chan values are pointers, and as such have equivalent memory layout. The structs you mention are what those pointers point to. My sample program above is just changing pointers.

Ouch! -- I thought they are represented as full data structures in SSA. Live and learn!

OK, it seems everyone agree that a pointer and a non-pointer can't alias; CL#632176 submitted. @randall77 , @ianlancetaylor , please kindly take a look when you'll have a spare moment.

Andrey
